File: apps/windows/TeslaSync.App/modals-dialogs/ChangelogModal/ChangelogModal.Model.cs
using System.Globalization;
using System.Text.RegularExpressions;
using TeslaSync.App.Core;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.ModalsDialogs;

/// <summary>
/// The Keep-a-Changelog change category (web <c>ChangelogChangeType</c>). Drives the section grouping,
/// the localized section heading and the section dot colour exactly like
/// web/src/components/feedback/ChangelogModal.tsx.
/// </summary>
public enum ChangelogChangeType
{
    /// <summary>New capability (web <c>added</c>).</summary>
    Added,

    /// <summary>Behaviour change (web <c>changed</c>).</summary>
    Changed,

    /// <summary>Bug fix (web <c>fixed</c>).</summary>
    Fixed,

    /// <summary>Removed capability (web <c>removed</c>).</summary>
    Removed,

    /// <summary>Deprecation notice (web <c>deprecated</c>).</summary>
    Deprecated,

    /// <summary>Security fix (web <c>security</c>).</summary>
    Security,
}

/// <summary>The release badge classification (web <c>ChangelogBadge</c>).</summary>
public enum ChangelogBadge
{
    /// <summary>The topmost release (web <c>latest</c>).</summary>
    Latest,

    /// <summary>A shipped stable release (web <c>stable</c>).</summary>
    Stable,

    /// <summary>A pre-release (web <c>beta</c>).</summary>
    Beta,
}

/// <summary>
/// The mutually-exclusive lifecycle state the <see cref="ChangelogModalViewModel"/> can be in — the native
/// union of the loading / loaded / empty / error / stale / offline branches a data-driven surface renders.
/// The web ChangelogModal reads static generated data so it only ever renders the loaded branch; the native
/// surface models the full cache-then-network seam so the catalog can be surfaced through a skeleton and a
/// degraded read still renders a friendly surface.
/// </summary>
public enum ChangelogModalState
{
    /// <summary>Initial fetch with no content yet — render the skeleton chrome.</summary>
    Loading,

    /// <summary>The catalog resolved with at least one release — render the body.</summary>
    Loaded,

    /// <summary>The catalog resolved empty — render the friendly empty surface.</summary>
    Empty,

    /// <summary>The read failed and nothing is renderable — render the retry affordance.</summary>
    Error,

    /// <summary>The shown body is backed by a cached read older than the freshness window — body plus a stale chip.</summary>
    Stale,

    /// <summary>The read is offline but cached content remains — body plus an offline chip.</summary>
    Offline,
}

/// <summary>A single typed change line within a release (web <c>ChangelogChange</c>).</summary>
/// <param name="Type">The Keep-a-Changelog category.</param>
/// <param name="Text">The human-readable change text (already localized at author time, shown verbatim).</param>
public sealed record ChangelogChange(ChangelogChangeType Type, string Text);

/// <summary>One release entry (web <c>ChangelogEntry</c>).</summary>
/// <param name="Version">Semver string, e.g. "0.7.0" or "1.0.0-beta.1".</param>
/// <param name="Date">ISO date (YYYY-MM-DD) the version was released.</param>
/// <param name="Badge">The badge classification.</param>
/// <param name="Changes">The flat list of changes, in section order.</param>
public sealed record ChangelogEntry(
    string Version,
    string Date,
    ChangelogBadge Badge,
    IReadOnlyList<ChangelogChange> Changes);

/// <summary>
/// The merged reading the <see cref="IChangelogSource"/> emits — the static catalog plus the user's
/// acknowledgement state (web <c>useChangelog</c> = generated <c>CHANGELOG</c> + the localStorage
/// <c>seen-version</c>). The projection derives the unseen subset and the first-visit flag from it.
/// </summary>
/// <param name="Entries">All releases, newest first.</param>
/// <param name="SeenVersion">The highest version the user has acknowledged, or <see langword="null"/> on a first visit.</param>
public sealed record ChangelogReading(IReadOnlyList<ChangelogEntry> Entries, string? SeenVersion)
{
    /// <summary>The topmost release version, or <see langword="null"/> when the catalog is empty.</summary>
    public string? LatestVersion => Entries.Count > 0 ? Entries[0].Version : null;
}

/// <summary>One section (added / changed / …) of a rendered release, with its localized heading and dot colour.</summary>
/// <param name="Type">The change category.</param>
/// <param name="Label">The localized section heading.</param>
/// <param name="DotStatus">The semantic status driving the leading dot colour (web <c>SECTION_DOT</c>).</param>
/// <param name="Items">The change texts in this section.</param>
public sealed record ChangelogSectionDisplay(
    ChangelogChangeType Type,
    string Label,
    StatusKind DotStatus,
    IReadOnlyList<string> Items);

/// <summary>A render-ready release entry: header metadata plus its non-empty, ordered sections.</summary>
/// <param name="Version">The raw semver string.</param>
/// <param name="VersionLabel">The display version ("v" + <paramref name="Version"/>).</param>
/// <param name="Date">The release date text.</param>
/// <param name="Badge">The badge classification.</param>
/// <param name="BadgeLabel">The localized badge label.</param>
/// <param name="BadgeStatus">The semantic status driving the badge colour (web <c>BADGE_VARIANT</c>).</param>
/// <param name="DefaultExpanded">True for the first two entries (web <c>defaultOpen={idx &lt; 2}</c>).</param>
/// <param name="Sections">The non-empty sections, in canonical order.</param>
/// <param name="AutomationName">The Narrator name for the entry header.</param>
public sealed record ChangelogEntryDisplay(
    string Version,
    string VersionLabel,
    string Date,
    ChangelogBadge Badge,
    string BadgeLabel,
    StatusKind BadgeStatus,
    bool DefaultExpanded,
    IReadOnlyList<ChangelogSectionDisplay> Sections,
    string AutomationName);

/// <summary>The complete, render-ready body the modal shows (web parity for the modal content + subtitle).</summary>
/// <param name="VisibleEntries">The entries shown — the unseen subset, or all entries on a first visit.</param>
/// <param name="IsFirstVisit">True when every entry is unseen (web <c>newEntries.length === entries.length</c>).</param>
/// <param name="HasUnseen">True when at least one release shipped after the seen version (web <c>hasUnseen</c>).</param>
/// <param name="NewCount">The count of unseen releases (drives the since-last-visit subtitle).</param>
/// <param name="Subtitle">The localized subtitle (first-visit copy or the since-last-visit count line).</param>
/// <param name="AutomationName">The Narrator name summarising the modal.</param>
public sealed record ChangelogModalDisplay(
    IReadOnlyList<ChangelogEntryDisplay> VisibleEntries,
    bool IsFirstVisit,
    bool HasUnseen,
    int NewCount,
    string Subtitle,
    string AutomationName)
{
    /// <summary>True when there is at least one entry to render.</summary>
    public bool HasEntries => VisibleEntries.Count > 0;
}

/// <summary>
/// The UI-free projection that turns a <see cref="ChangelogReading"/> into the render-ready
/// <see cref="ChangelogModalDisplay"/> — the native port of the body-building logic in
/// web/src/components/feedback/ChangelogModal.tsx (the unseen-subset gate, the first-visit subtitle, the
/// per-entry section grouping, the badge/section mapping) plus the semver comparison from
/// web/src/hooks/useChangelog.ts. Kept static and resource-free so it is exhaustively unit-testable without a
/// XAML runtime; every label flows through the injected <see cref="ILocalizer"/>.
/// </summary>
public static partial class ChangelogModalProjection
{
    private static readonly ChangelogChangeType[] SectionOrder =
    {
        ChangelogChangeType.Added,
        ChangelogChangeType.Changed,
        ChangelogChangeType.Fixed,
        ChangelogChangeType.Removed,
        ChangelogChangeType.Deprecated,
        ChangelogChangeType.Security,
    };

    /// <summary>
    /// Project the merged reading into the render-ready display: the unseen subset (or all entries on a first
    /// visit), the localized subtitle, and each entry's ordered non-empty sections.
    /// </summary>
    public static ChangelogModalDisplay Project(ChangelogReading reading, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(reading);
        ArgumentNullException.ThrowIfNull(localizer);

        IReadOnlyList<ChangelogEntry> entries = reading.Entries;
        string? seen = reading.SeenVersion;

        IReadOnlyList<ChangelogEntry> newEntries = string.IsNullOrEmpty(seen)
            ? entries
            : entries.Where(e => CompareVersions(e.Version, seen) > 0).ToList();

        bool isFirstVisit = newEntries.Count == entries.Count;
        IReadOnlyList<ChangelogEntry> visible = newEntries.Count > 0 ? newEntries : entries;

        string subtitle = isFirstVisit
            ? localizer.GetString(
                "changelog.modal.subtitleFirstVisit",
                "Welcome! Here's a quick tour of what TeslaSync ships with right now.")
            : string.Format(
                CultureInfo.CurrentCulture,
                localizer.GetString(
                    "changelog.modal.subtitleSinceLastVisit",
                    "{0} new release(s) since your last visit."),
                visible.Count);

        var entryDisplays = new List<ChangelogEntryDisplay>(visible.Count);
        for (int i = 0; i < visible.Count; i++)
        {
            entryDisplays.Add(ProjectEntry(visible[i], i, localizer));
        }

        string title = localizer.GetString("changelog.modal.title", "What's new in TeslaSync");
        string automationName = $"{title}. {subtitle}";

        return new ChangelogModalDisplay(
            entryDisplays,
            isFirstVisit,
            newEntries.Count > 0,
            newEntries.Count,
            subtitle,
            automationName);
    }

    /// <summary>The localized badge label for a badge (web <c>BADGE_KEY</c> / <c>BADGE_FALLBACK</c>).</summary>
    public static string BadgeLabel(ChangelogBadge badge, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        return badge switch
        {
            ChangelogBadge.Latest => localizer.GetString("changelog.badges.latest", "Latest"),
            ChangelogBadge.Stable => localizer.GetString("changelog.badges.stable", "Stable"),
            ChangelogBadge.Beta => localizer.GetString("changelog.badges.beta", "Beta"),
            _ => localizer.GetString("changelog.badges.stable", "Stable"),
        };
    }

    /// <summary>The semantic badge colour (web <c>BADGE_VARIANT</c>: latest→success, stable→info, beta→warning).</summary>
    public static StatusKind BadgeStatus(ChangelogBadge badge) => badge switch
    {
        ChangelogBadge.Latest => StatusKind.Success,
        ChangelogBadge.Stable => StatusKind.Info,
        ChangelogBadge.Beta => StatusKind.Warning,
        _ => StatusKind.Info,
    };

    /// <summary>The localized section heading (web <c>SECTION_KEY</c> / <c>SECTION_FALLBACK</c>).</summary>
    public static string SectionLabel(ChangelogChangeType type, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        return type switch
        {
            ChangelogChangeType.Added => localizer.GetString("changelog.sections.added", "Added"),
            ChangelogChangeType.Changed => localizer.GetString("changelog.sections.changed", "Changed"),
            ChangelogChangeType.Fixed => localizer.GetString("changelog.sections.fixed", "Fixed"),
            ChangelogChangeType.Removed => localizer.GetString("changelog.sections.removed", "Removed"),
            ChangelogChangeType.Deprecated => localizer.GetString("changelog.sections.deprecated", "Deprecated"),
            ChangelogChangeType.Security => localizer.GetString("changelog.sections.security", "Security"),
            _ => localizer.GetString("changelog.sections.changed", "Changed"),
        };
    }

    /// <summary>
    /// The section dot colour (web <c>SECTION_DOT</c>). The web palette has six tints; WinUI's
    /// <see cref="StatusKind"/> has five, so deprecated maps to neutral (the web purple has no semantic peer).
    /// </summary>
    public static StatusKind SectionDotStatus(ChangelogChangeType type) => type switch
    {
        ChangelogChangeType.Added => StatusKind.Success,
        ChangelogChangeType.Changed => StatusKind.Info,
        ChangelogChangeType.Fixed => StatusKind.Warning,
        ChangelogChangeType.Removed => StatusKind.Danger,
        ChangelogChangeType.Deprecated => StatusKind.Neutral,
        ChangelogChangeType.Security => StatusKind.Danger,
        _ => StatusKind.Neutral,
    };

    /// <summary>
    /// Compare two semver strings, returning -1 / 0 / 1 (web <c>compareVersions</c>). Pre-release tags sort
    /// before the corresponding release; anything that fails to parse falls back to an ordinal compare so a
    /// malformed entry never throws.
    /// </summary>
    public static int CompareVersions(string a, string b)
    {
        ArgumentNullException.ThrowIfNull(a);
        ArgumentNullException.ThrowIfNull(b);
        if (string.Equals(a, b, StringComparison.Ordinal))
        {
            return 0;
        }

        VersionParts? pa = Parse(a);
        VersionParts? pb = Parse(b);
        if (pa is not { } va || pb is not { } vb)
        {
            return Math.Sign(string.CompareOrdinal(a, b));
        }

        if (va.Major != vb.Major)
        {
            return va.Major < vb.Major ? -1 : 1;
        }

        if (va.Minor != vb.Minor)
        {
            return va.Minor < vb.Minor ? -1 : 1;
        }

        if (va.Patch != vb.Patch)
        {
            return va.Patch < vb.Patch ? -1 : 1;
        }

        // Cores equal — a pre-release sorts before the stable release.
        if (va.Pre is null && vb.Pre is not null)
        {
            return 1;
        }

        if (va.Pre is not null && vb.Pre is null)
        {
            return -1;
        }

        if (va.Pre is null && vb.Pre is null)
        {
            return 0;
        }

        return Math.Sign(string.CompareOrdinal(va.Pre, vb.Pre));
    }

    private static ChangelogEntryDisplay ProjectEntry(ChangelogEntry entry, int index, ILocalizer localizer)
    {
        var sections = new List<ChangelogSectionDisplay>(SectionOrder.Length);
        foreach (var type in SectionOrder)
        {
            var items = entry.Changes.Where(c => c.Type == type).Select(c => c.Text).ToList();
            if (items.Count == 0)
            {
                continue;
            }

            sections.Add(new ChangelogSectionDisplay(type, SectionLabel(type, localizer), SectionDotStatus(type), items));
        }

        string badgeLabel = BadgeLabel(entry.Badge, localizer);
        string versionLabel = $"v{entry.Version}";
        string automationName = $"{versionLabel}, {badgeLabel}, {entry.Date}";

        return new ChangelogEntryDisplay(
            entry.Version,
            versionLabel,
            entry.Date,
            entry.Badge,
            badgeLabel,
            BadgeStatus(entry.Badge),
            index < 2,
            sections,
            automationName);
    }

    private static VersionParts? Parse(string version)
    {
        var match = SemverRegex().Match(version);
        if (!match.Success)
        {
            return null;
        }

        return new VersionParts(
            int.Parse(match.Groups[1].Value, CultureInfo.InvariantCulture),
            int.Parse(match.Groups[2].Value, CultureInfo.InvariantCulture),
            int.Parse(match.Groups[3].Value, CultureInfo.InvariantCulture),
            match.Groups[4].Success ? match.Groups[4].Value : null);
    }

    [GeneratedRegex(@"^(\d+)\.(\d+)\.(\d+)(?:[-+](.+))?$")]
    private static partial Regex SemverRegex();

    private readonly record struct VersionParts(int Major, int Minor, int Patch, string? Pre);
}

File: apps/android/app/src/main/java/com/teslasync/modals-dialogs/ChangelogModal/ChangelogModalModel.kt
// Pure, framework-free model + projection for the ChangelogModal surface — the native analogue of the data
// the web component derives before returning JSX (web/src/components/feedback/ChangelogModal.tsx, backed by
// web/src/hooks/useChangelog.ts + web/src/generated/changelog.ts). No Compose, no Android, no HTTP: every
// type here is unit-tested off-device in the :android:testReleaseUnitTest gate, keeping the composable a
// thin render layer.
//
// The web ChangelogModal surfaces "what's new since last visit". Its data source, useChangelog, is NOT a
// network query — it reads a build-time-generated static catalog (@/generated/changelog) and a small
// acknowledgement record persisted in localStorage (seen-version + last-shown throttle), exposed through a
// useSyncExternalStore. That contract is reproduced here by [ChangelogModalModel] (the derivations:
// new-since-seen filtering, first-visit detection, the 24h auto-show throttle, Keep-a-Changelog section
// grouping, the default-open rule, and the seen/throttle reducers) over the [ChangelogSource] state-holder
// seam (P1/S8). Because the feed is static + local rather than remote, there is NO loading / error / stale /
// offline lifecycle to model — inventing one would be fabricated state the web source never has (a "No
// silent drift" covenant violation), exactly as the sibling AccordionSection surface documents. The
// genuinely reachable render states are the three [ChangelogRender] branches: FirstVisit (full history),
// SinceLastVisit (the delta), and Empty (a catalog with no releases — a friendly empty state, never a blank
// box).
//
// `InvalidPackageDeclaration` is suppressed: the mandated surface directory
// (com/teslasync/modals-dialogs/ChangelogModal — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path, exactly as the sibling feature-view surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.modalsdialogs.changelogmodal

/** Diagnostics + registry identifiers for the surface (P1/S11). */
object ChangelogRegistration {
    /** Stable surface id. */
    const val ID: String = "changelog-modal"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "ChangelogModal"
}

/** Keep-a-Changelog change category — the web `ChangelogChangeType` union. */
enum class ChangelogChangeType { Added, Changed, Fixed, Removed, Deprecated, Security }

/** UI badge classification for a release — the web `ChangelogBadge` union. */
enum class ChangelogBadge { Latest, Stable, Beta }

/** One typed change line within a release — the web `ChangelogChange`. */
data class ChangelogChange(
    val type: ChangelogChangeType,
    val text: String,
)

/** One released version with its typed changes — the web `ChangelogEntry`. */
data class ChangelogRelease(
    val version: String,
    val date: String,
    val badge: ChangelogBadge,
    val changes: List<ChangelogChange>,
)

/** A non-empty group of same-typed changes within a release (the web modal's per-section grouping). */
data class ChangelogSection(
    val type: ChangelogChangeType,
    val items: List<ChangelogChange>,
)

/**
 * The user's changelog acknowledgement record — the native analogue of the two localStorage keys the web
 * useChangelog persists: the highest [seenVersion] the user dismissed the modal for, and the [lastShownAt]
 * epoch-ms stamp that throttles the auto-show to once per 24h regardless of seen-state.
 */
data class ChangelogAck(
    val seenVersion: String? = null,
    val lastShownAt: Long? = null,
)

/**
 * The reduced render state of the modal body — the three mutually-exclusive, genuinely reachable branches.
 * Mirrors the web component's first-visit vs since-last-visit subtitle split, plus the native empty branch
 * (a catalog with no releases) that renders a friendly empty state instead of a blank box.
 */
sealed interface ChangelogRender {
    /** The catalog resolved to no releases — show a friendly empty state, never a blank box. */
    data object Empty : ChangelogRender

    /** First visit (seen-version null / all entries new) — the web `subtitleFirstVisit` path. */
    data class FirstVisit(
        val count: Int,
    ) : ChangelogRender

    /** Returning visit — [count] releases since the last visit (the web `subtitleSinceLastVisit` path). */
    data class SinceLastVisit(
        val count: Int,
    ) : ChangelogRender
}

/**
 * The localized strings the composable renders — resolved once at the render boundary and handed to the
 * stateless content as a framework-free bundle so the view stays a thin render layer. The modal title,
 * subtitles, action labels, badge labels, and section labels all resolve from the P1/S10 catalog; the
 * native-only empty hint + close affordance fall back to English defaults when no catalog entry exists.
 */
data class ChangelogStrings(
    val title: String,
    val viewFull: String,
    val gotIt: String,
    val closeLabel: String,
    val emptyMessage: String,
    val badgeLabels: Map<ChangelogBadge, String>,
    val sectionLabels: Map<ChangelogChangeType, String>,
) {
    /** The localized badge label for [badge], falling back to its enum name if unmapped. */
    fun badgeLabel(badge: ChangelogBadge): String = badgeLabels[badge] ?: badge.name

    /** The localized section heading for [type], falling back to its enum name if unmapped. */
    fun sectionLabel(type: ChangelogChangeType): String = sectionLabels[type] ?: type.name
}

/** Native-only microcopy defaults (no web catalog key ⇒ these English fallbacks are used). */
object ChangelogDefaults {
    /** Friendly empty-body hint shown when the catalog has no releases. */
    const val EMPTY_MESSAGE: String = "No release notes yet."

    /** Close affordance label for the modal (web relies on the DOM Esc/backdrop; TalkBack needs a label). */
    const val CLOSE_LABEL: String = "Close dialog"

    /** Action label announced for an entry header while collapsed (web relies on the DOM `aria-expanded`). */
    const val EXPAND_ACTION: String = "Expand"

    /** Action label announced for an entry header while expanded. */
    const val COLLAPSE_ACTION: String = "Collapse"

    /** State description announced while an entry is expanded — native equivalent of `aria-expanded="true"`. */
    const val EXPANDED_STATE: String = "Expanded"

    /** State description announced while an entry is collapsed — native equivalent of `aria-expanded="false"`. */
    const val COLLAPSED_STATE: String = "Collapsed"
}

/** Resource name for the native-only empty hint (by-name; absent ⇒ [ChangelogDefaults.EMPTY_MESSAGE]). */
const val KEY_EMPTY_MESSAGE: String = "translation_changelog_modal_empty"

/** Resource name for an entry's expand action label (by-name; absent ⇒ [ChangelogDefaults.EXPAND_ACTION]). */
const val KEY_EXPAND_ACTION: String = "translation_changelog_entry_expand"

/** Resource name for an entry's collapse action label (by-name; absent ⇒ [ChangelogDefaults.COLLAPSE_ACTION]). */
const val KEY_COLLAPSE_ACTION: String = "translation_changelog_entry_collapse"

/** Resource name for an entry's expanded state description (by-name; absent ⇒ default). */
const val KEY_EXPANDED_STATE: String = "translation_changelog_entry_expanded"

/** Resource name for an entry's collapsed state description (by-name; absent ⇒ default). */
const val KEY_COLLAPSED_STATE: String = "translation_changelog_entry_collapsed"

/**
 * The localized accessibility affordance strings for a collapsible entry header — resolved once at the
 * render boundary so the header's TalkBack action + state description track the open/closed toggle the way
 * the web `aria-expanded` does. The web source owns no text keys for these (it relies on the DOM), so they
 * resolve by-name with the English fallbacks above.
 */
data class ChangelogEntryAffordances(
    val expandAction: String,
    val collapseAction: String,
    val expandedState: String,
    val collapsedState: String,
) {
    /** The TalkBack action label for the toggle in its current [expanded] state. */
    fun actionLabel(expanded: Boolean): String = if (expanded) collapseAction else expandAction

    /** The TalkBack state description for the current [expanded] state (web `aria-expanded`). */
    fun stateLabel(expanded: Boolean): String = if (expanded) expandedState else collapsedState
}

/**
 * The state-holder seam the surface binds to — the native analogue of the web `useChangelog` hook (P1/S8).
 * The view performs no persistence or HTTP itself; it reads the catalog + acknowledgement through this port
 * and writes seen/throttle stamps back through it. A concrete implementation ([DefaultChangelogSource])
 * binds the embedded [ChangelogCatalog] to a platform-backed [ChangelogAckStore].
 */
interface ChangelogSource {
    /** All releases, newest first — the web `entries` / generated `CHANGELOG`. */
    val releases: List<ChangelogRelease>

    /** The highest released version — the web `latestVersion` / generated `LATEST_VERSION`. */
    val latestVersion: String

    /** Whether the user has finished onboarding (web `localStorage['teslasync-onboarded']`). */
    val hasCompletedOnboarding: Boolean

    /** The current acknowledgement record (seen-version + last-shown throttle stamp). */
    fun ack(): ChangelogAck

    /** Marks [latestVersion] as seen and stamps the throttle — the web `markSeen()`. */
    fun markSeen()

    /** Stamps the auto-show throttle WITHOUT marking seen — the web `stampShown()` (manual open). */
    fun stampShown()

    /** Wall-clock epoch-ms seam, injectable for tests — the web `Date.now()`. */
    fun now(): Long
}

/** Persistence seam for the acknowledgement record (a thin platform store backs it in production). */
interface ChangelogAckStore {
    /** Reads the persisted acknowledgement, or a blank record on first run. */
    fun read(): ChangelogAck

    /** Persists [ack] (seen-version + last-shown). */
    fun write(ack: ChangelogAck)
}

/**
 * The default [ChangelogSource]: the embedded [ChangelogCatalog] over a platform-backed [ChangelogAckStore],
 * with the seen/throttle reducers delegated to the pure [ChangelogModalModel]. Framework-free (the clock +
 * onboarding probe are injectable seams), so the binding logic is exercised off-device.
 */
class DefaultChangelogSource(
    private val store: ChangelogAckStore,
    private val onboardingProbe: () -> Boolean,
    override val releases: List<ChangelogRelease> = ChangelogCatalog.releases,
    override val latestVersion: String = ChangelogCatalog.LATEST_VERSION,
    private val clock: () -> Long = { System.currentTimeMillis() },
) : ChangelogSource {
    override val hasCompletedOnboarding: Boolean get() = onboardingProbe()

    override fun ack(): ChangelogAck = store.read()

    override fun markSeen() {
        store.write(ChangelogModalModel.markSeen(store.read(), latestVersion, now()))
    }

    override fun stampShown() {
        store.write(ChangelogModalModel.stampShown(store.read(), now()))
    }

    override fun now(): Long = clock()
}

/**
 * Pure, side-effect-free reducer — the native port of useChangelog's derivations + the web component's
 * prop-and-state to render mapping. Stateless so it is fully covered by the off-device unit gate.
 */
object ChangelogModalModel {
    /** Auto-show throttle window — the web `AUTO_SHOW_THROTTLE_MS` (once per 24h). */
    const val AUTO_SHOW_THROTTLE_MS: Long = 24L * 60L * 60L * 1000L

    /** Settle delay before the auto-show eligibility re-check fires — the web `AUTO_SHOW_DELAY_MS`. */
    const val AUTO_SHOW_DELAY_MS: Long = 2_000L

    /** Number of leading entries expanded by default — the web `defaultOpen={idx < 2}`. */
    const val DEFAULT_OPEN_COUNT: Int = 2

    /** External "view full changelog" target — the web `github.com/.../releases` link. */
    const val RELEASES_URL: String = "https://github.com/ev-dev-labs/teslasync/releases"

    /** Canonical section order — the web `SECTION_ORDER`. */
    val SECTION_ORDER: List<ChangelogChangeType> =
        listOf(
            ChangelogChangeType.Added,
            ChangelogChangeType.Changed,
            ChangelogChangeType.Fixed,
            ChangelogChangeType.Removed,
            ChangelogChangeType.Deprecated,
            ChangelogChangeType.Security,
        )

    /**
     * Compares two semver strings the way the web `compareVersions` does: numeric MAJOR.MINOR.PATCH, with
     * pre-release tags (`-beta.N`) sorting BEFORE the corresponding release, and a lexicographic fallback
     * for anything that fails to parse. Returns -1, 0, or 1.
     */
    fun compareVersions(
        a: String,
        b: String,
    ): Int {
        if (a == b) return 0
        val pa = parseSemver(a)
        val pb = parseSemver(b)
        return if (pa == null || pb == null) sign(a.compareTo(b)) else compareSemver(pa, pb)
    }

    /**
     * Releases that shipped after [seenVersion] — the web `newEntries`. A null seen-version (first visit)
     * yields the entire history, which is also the right onboarding behaviour.
     */
    fun newReleases(
        all: List<ChangelogRelease>,
        seenVersion: String?,
    ): List<ChangelogRelease> =
        if (seenVersion == null) {
            all
        } else {
            all.filter { compareVersions(it.version, seenVersion) > 0 }
        }

    /** Whether there are any unseen releases — the web `hasUnseen`. */
    fun hasUnseen(newReleases: List<ChangelogRelease>): Boolean = newReleases.isNotEmpty()

    /**
     * The list shown inside the modal — the unseen subset when there is one, otherwise the full history
     * (web `visibleEntries = newEntries.length > 0 ? newEntries : entries`).
     */
    fun visibleReleases(
        newReleases: List<ChangelogRelease>,
        all: List<ChangelogRelease>,
    ): List<ChangelogRelease> = if (newReleases.isNotEmpty()) newReleases else all

    /** First visit when every release is new — the web `isFirstVisit = newEntries.length === entries.length`. */
    fun isFirstVisit(
        newCount: Int,
        totalCount: Int,
    ): Boolean = newCount == totalCount

    /**
     * Whether enough time has passed since the last auto-show — the web `canAutoShow`. False when there is
     * nothing unseen; true on the first eligible run; otherwise gated by the 24h throttle window.
     */
    fun canAutoShow(
        hasUnseen: Boolean,
        lastShownAt: Long?,
        nowMs: Long,
        throttleMs: Long = AUTO_SHOW_THROTTLE_MS,
    ): Boolean =
        when {
            !hasUnseen -> false
            lastShownAt == null -> true
            else -> nowMs - lastShownAt >= throttleMs
        }

    /**
     * The full auto-show gating predicate — the web effect's guard (`hasUnseen && hasCompletedOnboarding &&
     * canAutoShow`), plus a [suppressed] seam for the native equivalent of the `[data-tour-active]` probe so
     * the modal never stacks on top of an active tour/onboarding overlay.
     */
    fun shouldAutoShow(
        hasUnseen: Boolean,
        hasCompletedOnboarding: Boolean,
        canAutoShow: Boolean,
        suppressed: Boolean,
    ): Boolean = hasUnseen && hasCompletedOnboarding && canAutoShow && !suppressed

    /**
     * Groups a release's changes by canonical type, dropping empty sections — the web modal's per-entry
     * `SECTION_ORDER.map(...).filter(g => g.items.length > 0)`.
     */
    fun groupChanges(changes: List<ChangelogChange>): List<ChangelogSection> =
        SECTION_ORDER.mapNotNull { type ->
            val items = changes.filter { it.type == type }
            if (items.isEmpty()) null else ChangelogSection(type, items)
        }

    /** Whether the entry at [index] starts expanded — the web `defaultOpen={idx < 2}`. */
    fun defaultExpanded(index: Int): Boolean = index < DEFAULT_OPEN_COUNT

    /** Marks [latestVersion] as seen and stamps the throttle at [nowMs] — the web `markSeen()`. */
    fun markSeen(
        ack: ChangelogAck,
        latestVersion: String,
        nowMs: Long,
    ): ChangelogAck = ack.copy(seenVersion = latestVersion, lastShownAt = nowMs)

    /** Stamps the throttle at [nowMs] without marking seen — the web `stampShown()`. */
    fun stampShown(
        ack: ChangelogAck,
        nowMs: Long,
    ): ChangelogAck = ack.copy(lastShownAt = nowMs)

    /**
     * Classifies the modal body into its [ChangelogRender] state from the visible-entry count and the
     * first-visit flag. An empty catalog renders the friendly empty state instead of a blank box.
     */
    fun classify(
        visibleCount: Int,
        isFirstVisit: Boolean,
    ): ChangelogRender =
        when {
            visibleCount == 0 -> ChangelogRender.Empty
            isFirstVisit -> ChangelogRender.FirstVisit(visibleCount)
            else -> ChangelogRender.SinceLastVisit(visibleCount)
        }

    // ── Internal semver helpers (web `compareVersions`) ──────────────────────────

    private val SEMVER = Regex("""^(\d+)\.(\d+)\.(\d+)(?:[-+](.+))?$""")

    private fun parseSemver(value: String): Semver? {
        val match = SEMVER.matchEntire(value) ?: return null
        val groups = match.groupValues
        val core = listOf(groups[1].toInt(), groups[2].toInt(), groups[3].toInt())
        return Semver(core, groups[4].ifEmpty { null })
    }

    private fun compareSemver(
        a: Semver,
        b: Semver,
    ): Int {
        val coreDiff =
            (0..2)
                .asSequence()
                .map { sign(a.core[it].compareTo(b.core[it])) }
                .firstOrNull { it != 0 }
        return coreDiff ?: comparePre(a.pre, b.pre)
    }

    private fun comparePre(
        a: String?,
        b: String?,
    ): Int =
        when {
            a == null && b == null -> 0
            a == null -> 1
            b == null -> -1
            else -> sign(a.compareTo(b))
        }

    private fun sign(value: Int): Int = value.coerceIn(-1, 1)

    private data class Semver(
        val core: List<Int>,
        val pre: String?,
    )
}

/**
 * Reproduces i18next's `t(key, default)` against the native i18n facade: returns the [lookup] result for
 * [resourceName] when it resolves to a non-blank string, otherwise the [fallback] default. [lookup] is a
 * thin seam over the Android string catalog in production (an optional by-name resource read) and a map in
 * tests, so the resolve-or-fallback decision stays pure and unit-tested.
 */
fun resolveOptional(
    lookup: (String) -> String?,
    resourceName: String,
    fallback: String,
): String = lookup(resourceName)?.takeIf { it.isNotBlank() } ?: fallback
